File: lib/cc/importer/course_settings.rb
#
# Copyright (C) 2011 Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#
module CC::Importer
  module CourseSettings
    include CC::Importer
    
    def course_settings_doc(file)
      open_file_xml File.join(@unzipped_file_path, COURSE_SETTINGS_DIR, file)
    end
    
    def convert_non_dependant_course_settings
      @course[:course] = convert_course_settings(course_settings_doc(COURSE_SETTINGS))
      @course[:assignment_groups] = convert_assignment_groups(course_settings_doc(ASSIGNMENT_GROUPS))
      @course[:external_tools] = convert_external_tools(course_settings_doc(EXTERNAL_TOOLS))
    end

    def convert_course_settings(doc)
      course = {}
      return course unless doc
      course[:migration_id] = get_node_att(doc, 'course',  'identifier')

      ['title', 'course_code', 'hashtag', 'default_wiki_editing_roles',
       'turnitin_comments', 'default_view', 'license',
       'group_weighting_scheme'].each do |string_type|
        val = get_node_val(doc, string_type)
        course[string_type] = val unless val.nil?
      end
      ['is_public', 'indexed', 'publish_grades_immediately', 'allow_student_wiki_edits',
       'allow_student_assignment_edits', 'show_public_context_messages',
       'allow_student_forum_attachments', 'allow_student_organized_groups',
       'show_all_discussion_entries', 'open_enrollment', 'allow_wiki_comments',
       'self_enrollment'].each do |bool_val|
        val = get_bool_val(doc, bool_val)
        course[bool_val] = val unless val.nil?
      end
      ['start_at', 'conclude_at'].each do |date_type|
        val = get_time_val(doc, date_type)
        course[date_type] = val unless val.nil?
      end

      course['storage_quota'] = get_int_val(doc, 'storage_quota')
      
      course
    end

    def convert_assignment_groups(doc = nil)
      groups = []
      return groups unless doc
      doc.css('assignmentGroup').each do |node|
        group = {}
        group['migration_id'] = node['identifier']
        group['title'] = get_node_val(node, 'title')
        group['position'] = get_int_val(node, 'position')
        group['group_weight'] = get_float_val(node, 'group_weight')
        group['rules'] = []
        node.css('rules rule').each do |r_node|
          rule = {}
          rule['drop_type'] = get_node_val(r_node, 'drop_type')
          rule['drop_count'] = get_int_val(r_node, 'drop_count')
          rule['assignment_migration_id'] = get_node_val(r_node, 'identifierref')
          group['rules'] << rule
        end
        
        groups << group
      end
      
      groups
    end

    def convert_external_tools(doc)
      tools = []
      return tools unless doc
      doc.css('externalTool').each do |node|
        tool = {}
        tool['migration_id'] = node['identifier']
        tool['title'] = get_node_val(node, 'title')
        tool['description'] = get_node_val(node, 'description')
        tool['domain'] = get_node_val(node, 'domain')
        tool['url'] = get_node_val(node, 'url')
        tool['privacy_level'] = get_node_val(node, 'privacy_level')
        
        tools << tool
      end

      tools
    end

  end
end

File: app/jsx/bundles/course_settings.js
/*
 * Copyright (C) 2011 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import React from 'react'
import ReactDOM from 'react-dom'
import NavigationView from 'compiled/views/course_settings/NavigationView'
import FeatureFlagAdminView from 'compiled/views/feature_flags/FeatureFlagAdminView'
import CourseImageSelector from '../course_settings/components/CourseImageSelector'
import BlueprintLockOptions from '../blueprint_courses/components/BlueprintLockOptions'
import CourseAvailabilityOptions from '../course_settings/components/CourseAvailabilityOptions'
import configureStore from '../course_settings/store/configureStore'
import initialState from '../course_settings/store/initialState'
import 'course_settings'
import 'grading_standards'
import FeatureFlags from '../feature_flags/FeatureFlags'

const blueprint = document.getElementById('blueprint_menu')
if (blueprint) {
  ReactDOM.render(
    <BlueprintLockOptions
      isMasterCourse={ENV.IS_MASTER_COURSE}
      disabledMessage={ENV.DISABLED_BLUEPRINT_MESSAGE}
      generalRestrictions={ENV.BLUEPRINT_RESTRICTIONS}
      useRestrictionsbyType={ENV.USE_BLUEPRINT_RESTRICTIONS_BY_OBJECT_TYPE}
      restrictionsByType={ENV.BLUEPRINT_RESTRICTIONS_BY_OBJECT_TYPE}
    />,
    blueprint
  )
}

const navView = new NavigationView({el: $('#tab-navigation')})

if (document.getElementById('tab-features')) {
  if (window.ENV.NEW_FEATURES_UI) {
    ReactDOM.render(<FeatureFlags disableDefaults />, document.getElementById('tab-features'))
  } else {
    const featureFlagView = new FeatureFlagAdminView({el: '#tab-features'})
    featureFlagView.collection.fetchAll()
  }
}

$(() => navView.render())

if (ENV.COURSE_IMAGES_ENABLED) {
  const courseImageStore = configureStore(initialState)

  ReactDOM.render(
    <CourseImageSelector store={courseImageStore} name="course[image]" courseId={ENV.COURSE_ID} />,
    $('.CourseImageSelector__Container')[0]
  )
}

const availabilityOptionsContainer = document.getElementById('availability_options_container')
if (availabilityOptionsContainer) {
  ReactDOM.render(
    <CourseAvailabilityOptions
      canManage={
        ENV.PERMISSIONS.manage_courses ||
        (ENV.PERMISSIONS.manage && !ENV.PREVENT_COURSE_AVAILABILITY_EDITING_BY_TEACHERS)
      }
      viewPastLocked={ENV.RESTRICT_STUDENT_PAST_VIEW_LOCKED}
      viewFutureLocked={ENV.RESTRICT_STUDENT_FUTURE_VIEW_LOCKED}
    />,
    availabilityOptionsContainer
  )
}
